### PR TITLE
Make arbitrary mipmap detection a config option

### PR DESCRIPTION
**Pull request by JonnyH**
Under GFX::Enhancements::ArbitraryMipmapDetection - default enabled

### CODE REVIEW

**Comment by legoj15** [comment]
IMO weight should be an option. There are cases where arb mipmaps are detected in one part of a game, but not other parts

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 1]
```diff
@@ -605,10 +608,13 @@ class ArbitraryMipmapDetector
...
         for (u32 j = 0; j < shape.width; ++j, row1 += 4, row2 += 4)
         {
-          average_diff += std::abs(static_cast<float>(row1[0]) - static_cast<float>(row2[0]));
-          average_diff += std::abs(static_cast<float>(row1[1]) - static_cast<float>(row2[1]));
-          average_diff += std::abs(static_cast<float>(row1[2]) - static_cast<float>(row2[2]));
-          average_diff += std::abs(static_cast<float>(row1[3]) - static_cast<float>(row2[3]));
+          for (int channel = 0; channel < 4; channel++)
+          {
+            const float diff =
+                std::abs(static_cast<float>(row1[channel]) - static_cast<float>(row2[channel]));
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: There is no need to call std::abs here, neither for a cast to float.
int diff = row1[channel] - row2[channel];
int diff_squared  = diff * diff;

But indeed, the sum may overflow.

Comment (JonnyH) [reply Source/Core/VideoCommon/TextureCacheBase.cpp]: Generally I feel like summing this in float space isn't really needed, but 8bit*8bit * 4 channels would need 18bits of precision, so in a 32bit uint it'll only take a 128x128 square texture of max difference to overflow. Pushing to 64 bit would solve this (and likely be the best solution, as it'll only ever be 'slow' on platforms that don't have JITs anyway?), or scale before the sum (which may cause 'small' differences to be truncated to zero)

**Comment by JMC47** [comment]
The option does work at disabling Arbitrary Mipmaps.  Wind Waker's sea effect and Super Mario Galaxy's lava effects can be broken again.

And I guess some textures are slightly sharper in some games at high distances.

**Pull request merged**